### ISSUE TITLE
Incorrect and inconsistent error for mismatching byref arguments' escape scopes

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#27357]
**Version Used**: 

Microsoft Visual Studio Professional 2017 
Version 15.7.2
VisualStudio.15.Release/15.7.2+27703.2018
Microsoft .NET Framework
Version 4.7.03056

**Steps to Reproduce**:

```cs
using System;

static class C
{
  static void M(Span<int> s1)
  {
    Span<int> s2 = stackalloc int[10];
    M(s2, s1); // no errors here
    M(s2, in s1); // errors
  }

  public static void M(Span<int> x, in Span<int> y)
  {
  }
}
```

**Expected Behavior**:
Either errors for both invocations or for neither of them

**Actual Behavior**:
Errors reported only for the second invocation.
```
Error CS8350: This combination of arguments to 'C.M(Span<int>, in Span<int>)' is disallowed because it may expose variables referenced by parameter 'x' outside of their declaration scope
Error CS8352: Cannot use local 's2' in this context because it may expose referenced variables outside of their declaration scope
```

**Notes**
These errors should be reported either for both invocations or for neither of them since these calls are equivalent.
I personally think that this error should not be reported for `in` arguments at all because possible byvalue leaks to `ref`/`out` arguments are already reported regardless of `in` arguments and their byref assignments will be limited to the callee.

**Comment by TessenR** [comment]
Same problem here:
```cs
using System;

class Program
{
  static void Main()
  {
  }
 
  void Test1()
  {
    S1 rOuter = default;
    rOuter.field = default;
 
    Span<int> inner = stackalloc int[1];
    S1 rInner = MayWrap(inner);
 
    int dummy1 = rInner[rOuter]; // valid
    int dummy2 = rInner[in rOuter]; // Error CS8352: Cannot use local 'rInner' in this context because it may expose referenced variables outside of their declaration scope

  }

  static extern S1 MayWrap(in Span<int> arg);
 
  ref struct S1
  {
    public Span<int> field;
    public extern int this[in S1 arg1] { get; }
  }
}
```

**Comment by TessenR** [comment]
Similar problem exists with parameters' ref escape scopes:
```cs
class C
{
  static ref Span<int> M1(ref Span<int> x)
  {
    ref Span<int> q = ref x;
    return ref q; // error cannot be returned by reference
  }

  static ref Span<int> M2(ref Span<int> x)
  {
    return ref x; // OK
  }
}
```

In this case it's caused by discrepancy between [GetRefEscape](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Binder/Binder.ValueChecks.cs,604613d82bb46358) and  [CheckParameterRefEscape](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Binder/Binder.ValueChecks.cs,376ab6d77b1940ea) methods' logic

`GetRefEscape` defines `ref` parameters' escape scope as `TopLevelScope`
```cs
                    // byval parameters can escape to method's top level.
                    // others can be escape further, unless they are ref-like.
                    // NOTE: "method" here means nearest containing method, lambda or nested method
                    return parameter.RefKind == RefKind.None || parameter.Type?.IsByRefLikeType == true ?
                        Binder.TopLevelScope :
                        Binder.ExternalScope;
```

`CheckParameterRefEscape` on the other hand allows `ref` parameters to escape to any scope:
```cs
        private static bool CheckParameterRefEscape(SyntaxNode node, BoundParameter parameter, uint escapeTo, bool checkingReceiver, DiagnosticBag diagnostics)
        {
            ParameterSymbol parameterSymbol = parameter.ParameterSymbol;
 
            // byval parameters can escape to method's top level.
            // others can be escape further, unless they are ref-like.
            if (escapeTo == Binder.ExternalScope && parameterSymbol.RefKind == RefKind.None)
            {
                if (checkingReceiver)
                {
                    Error(diagnostics, ErrorCode.ERR_RefReturnParameter2, parameter.Syntax, parameterSymbol.Name);
                }
                else
                {
                    Error(diagnostics, ErrorCode.ERR_RefReturnParameter, node, parameterSymbol.Name);
                }
                return false;
            }
 
            // can ref-escape to any scope otherwise
            return true;
        }
```

I believe that these two methods were intended to have similar logic and have consistent results.

**Comment by OmarTawfik** [comment]
There are actually two separate issues here:
* Inconsistency of errors because the argument RefKind is being checked instead of the parameter (should be tracked by this)
* Returning ref parameters by ref for ref-like types (filed #27874 for it).
